plugin/storage/badger/factory.go

```diff
@@ -0,0 +1,179 @@
+type Factory struct {
+	Options *Options
+	store   *badger.DB
+	cache   *badgerStore.CacheStore
+	logger  *zap.Logger
+
+	tmpDir              string
+	maintenanceInterval time.Duration
+	maintenanceTicker   *time.Ticker
+}
+
+const (
+	defaultTickerInterval time.Duration = 5 * time.Minute
+)
+
+// NewFactory creates a new Factory.
+func NewFactory() *Factory {
+	if ValueLogSpaceAvailable == nil {
+		ValueLogSpaceAvailable = expvar.NewInt("badger_value_log_bytes_available")
+	}
+	if KeyLogSpaceAvailable == nil {
+		KeyLogSpaceAvailable = expvar.NewInt("badger_key_log_bytes_available")
+	}
+	if LastMaintenanceRun == nil {
+		LastMaintenanceRun = expvar.NewInt("badger_storage_maintenance_last_run")
+	}
+	if LastValueLogCleaned == nil {
+		LastValueLogCleaned = expvar.NewInt("badger_storage_valueloggc_last_run")
+	}
+	return &Factory{
+		Options:             NewOptions("badger"),
+		maintenanceInterval: defaultTickerInterval,
+	}
+}
+
+// AddFlags implements plugin.Configurable
+func (f *Factory) AddFlags(flagSet *flag.FlagSet) {
+	f.Options.AddFlags(flagSet)
+}
+
+// InitFromViper implements plugin.Configurable
+func (f *Factory) InitFromViper(v *viper.Viper) {
+	f.Options.InitFromViper(v)
+}
+
+// Initialize implements storage.Factory
+func (f *Factory) Initialize(metricsFactory metrics.Factory, logger *zap.Logger) error {
+	f.logger = logger
+
+	opts := badger.DefaultOptions
+
+	if f.Options.primary.Ephemeral {
+		opts.SyncWrites = false
+		// Error from TempDir is ignored to satisfy Codecov
+		dir, _ := ioutil.TempDir("", "badger")
+		f.tmpDir = dir
+		opts.Dir = f.tmpDir
+		opts.ValueDir = f.tmpDir
+
+		f.Options.primary.KeyDirectory = f.tmpDir
+		f.Options.primary.ValueDirectory = f.tmpDir
+	} else {
+		opts.SyncWrites = f.Options.primary.SyncWrites
+		opts.Dir = f.Options.primary.KeyDirectory
+		opts.ValueDir = f.Options.primary.ValueDirectory
+	}
+
+	store, err := badger.Open(opts)
+	if err != nil {
+		return err
+	}
+	f.store = store
+
+	// Error ignored to satisfy Codecov
+	cache, _ := badgerStore.NewCacheStore(f.store, f.Options.primary.SpanStoreTTL)
+	f.cache = cache
+
+	f.maintenanceTicker = time.NewTicker(f.maintenanceInterval)
+	go f.maintenance()
+
+	logger.Info("Badger storage configuration", zap.Any("configuration", opts))
+
+	return nil
+}
+
+// CreateSpanReader implements storage.Factory
+func (f *Factory) CreateSpanReader() (spanstore.Reader, error) {
+	return badgerStore.NewTraceReader(f.store, f.cache), nil
+}
+
+// CreateSpanWriter implements storage.Factory
+func (f *Factory) CreateSpanWriter() (spanstore.Writer, error) {
+	return badgerStore.NewSpanWriter(f.store, f.cache, f.Options.primary.SpanStoreTTL, f), nil
+}
+
+// CreateDependencyReader implements storage.Factory
+func (f *Factory) CreateDependencyReader() (dependencystore.Reader, error) {
+	return nil, nil
+}
+
+// Close Implements io.Closer and closes the underlying storage
+func (f *Factory) Close() error {
+	f.maintenanceTicker.Stop()
```

Comment:
this won't make the `maintenance()` func exit since the ticker channel is never closed.